### ISSUE TITLE
The testing on the A800 and H800 based examples shows no effect, and the training actually slows down significantly.

### ISSUE THREAD

**Issue by TheQziEDGnb** [Azure/MS-AMP#198]
Using MS-AMP should be able to significantly improve both training and inference, but now the speed is slower，i don't know if I'm using it incorrectly. 

![Image](https://github.com/user-attachments/assets/dd5c5eb6-0335-4964-ad98-581390415243)

**How to reproduce it?**:

device : A800 and H800

env : Use the docker method mentioned in the document [There](https://azure.github.io/MS-AMP/docs/getting-started/installation)

Then do not modify any parameter，only add a timing code at mnist.py：
```
    train_time = []
    test_time = []
    for epoch in range(1, args.epochs + 1):
        start = time.time()
        train(args, model, device, train_loader, optimizer, epoch)
        train_time.append(time.time() - start)
        mid = time.time()
        test(model, device, test_loader)
        test_time.append(time.time() - mid)
        scheduler.step()

    print("train time avg : " + str(sum(train_time) / len(train_time)))
    print("test time avg : " + str(sum(test_time) / len(test_time)))
```

command

```
python mnist.py --enable-msamp --opt-level=O2
python mnist.py --enable-msamp --opt-level=O1
python mnist.py
```

**Comment by Mr-Philo** [comment]
Hi @TheQziEDGnb , thanks for your issue!

The speed you measured is normal and as expected. The reasons are:

1. MS-AMP is specially designed for accelerating Transformer architecture, where **matrix multiplication**  takes up most of its computations. Under the Mnist example, the model structure use **convolutional computations** a lot, thus the speed boosting effect may be small.
```py
class Net(nn.Module):
    """The neural network model for mnist."""
    def __init__(self):
        """Constructor."""
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.25)
        self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)
```
Note that current ViT (Vision Transformer) architecture will have a much better performance gain under MS-AMP than traditional CNNs.

2. MS-AMP's effect of speed acceleration **is not significant in small models**, because **it needs extra precision casting between FP8 and FP32(or FP16)**. Under our previous experiments, it may be better to evaluate the speed performance on Transformers larger than 1B parameters.
If you want to explore speed performance in more depth, we suggest you refer to the library [MS-AMP-Examples](https://github.com/Azure/MS-AMP-Examples), where the running scripts of larger models are provided. 😊

For more discussion related to the speed performance, please see #172, #175

**Comment by Mr-Philo** [comment]
Close as completed. Please feel free to reopen it for further discussion.

**Comment by TheQziEDGnb** [comment]
Thanks for your reply very much! When I use MS-AMP in my small llama architecture model, it still doesn't work.

My code:
```
def train(iter_num):
    scale = config.dim_model_base / config.hidden_size
    lr = get_lr(iter_num) if decay_lr else learning_rate
    for param_group in optimizer.param_groups:
        param = param_group['params'][0]
        if param.dim() >= 2:
            # Adjust the learning rate of each two-dimensional tensor parameter
            param_group['lr'] = lr * scale
        else:
            param_group['lr'] = lr

    batch = get_batch('train')

    # Zero your gradients for every batch!
    optimizer.zero_grad()

    # Compute the loss and its gradients
    loss = model(batch, labels=batch).loss
    print(loss.dtype)
    loss.backward()

    # Adjust learning weights
    optimizer.step()

    if iter_num % log_interval == 0:
        lossf = loss.item()
        bar_desc = f"loss: {lossf:.4f}"
        t.set_description(bar_desc, refresh=True)
```

```
from msamp.optim import FSDPAdam

    optimizer = optim.AdamW(model.parameters(), lr=learning_rate)

    if args.enable_msamp:
        print(f'msamp is enabled, opt_level: {args.opt_level}')
        import msamp
        model, optimizer = msamp.initialize(model, optimizer, opt_level=args.opt_level)

    end_iter = total_steps + 1
    scope = range(start_iter, end_iter)
    t = tqdm(scope, initial=start_iter, total=end_iter, desc='', leave=True)

    for iter_num in t:
        train(iter_num)
```
The memory usage of O2 optimized training is larger than that of unoptimized training, and when I output the dtype of loss in train, it shows fp32.

I think there may be something wrong with my usage, please help me correct it，thanks！

**Comment by TheQziEDGnb** [comment]
Thanks for your reply very much！I have a new question in the comments of
the issue.

Ruizhe Wang ***@***.***> 于2025年1月7日周二 16:03写道：

**Comment by TheQziEDGnb** [comment]
Thanks for your reply very much！I have a new question in the comments of
the issue.

Ruizhe Wang ***@***.***> 于2025年1月7日周二 16:03写道：
